includes/data-stores/class-wc-product-data-store-cpt.php

```diff
@@ -1414,6 +1414,39 @@ public function update_product_stock( $product_id_with_stock, $stock_quantity =
 		return $new_stock;
 	}
 
+	/**
+	 * Update a product's sale count based on query for previous orders.
+	 *
+	 * @since  4.3
+	 * @param  int        $product_id Product ID.
+	 * @param  null|array $exclude_order_ids An array of order IDs to exclude from the count.
+	 */
+	public function update_product_sales_by_query( $product_id, $exclude_order_ids = null ) {
+		global $wpdb;
+
+		$prepare_args = array();
+		$exclude_order_id_clause = '';
+		if ( is_array( $exclude_order_ids ) && ! empty( $exclude_order_ids ) ) {
+			$prepare_args = $exclude_order_ids;
+			$exclude_order_id_clause = 'posts.ID NOT IN ( ' . implode( ',', array_fill( 0, count( $exclude_order_ids ), '%d' ) ) . ') AND ';
+		}
+
+		$query = "SELECT SUM( order_item_meta__qty.meta_value ) AS order_item_count FROM {$wpdb->posts} AS posts " .
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_items AS order_items ON posts.ID = order_items.order_id " .
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS order_item_meta__qty ON " .
+			"order_items.order_item_id = order_item_meta__qty.order_item_id AND order_item_meta__qty.meta_key = '_qty' " .
+			"INNER JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS order_item_meta__product_id_array ON " .
+			'order_items.order_item_id = order_item_meta__product_id_array.order_item_id ' .
+			"WHERE {$exclude_order_id_clause} posts.post_type IN ( 'shop_order','shop_order_refund' ) " .
+			"AND posts.post_status IN ( 'wc-completed', 'wc-processing', 'wc-on-hold' ) " .
```

Comment:
I refunded half of the the test orders product count. The updated `total_sales` was `0.0`. Does `wc-refunded` need to be included in this list to balance the negative amount in the refund order?

Reply:
Good catch. Thanks. I updated the query and added a new test to handle refunds.